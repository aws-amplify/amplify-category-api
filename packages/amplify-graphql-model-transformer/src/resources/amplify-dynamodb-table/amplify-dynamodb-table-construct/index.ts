import { aws_kms as kms, CustomResource, Lazy, Resource } from 'aws-cdk-lib';
import {
  Attribute,
  BillingMode,
  CfnTable,
  CfnTableProps,
  GlobalSecondaryIndexProps,
  ITable,
  LocalSecondaryIndexProps,
  ProjectionType,
  SchemaOptions,
  SecondaryIndexProps,
  Table,
  TableEncryption,
  TableProps,
} from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

const HASH_KEY_TYPE = 'HASH';
const RANGE_KEY_TYPE = 'RANGE';

// https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#limits-secondary-indexes
const MAX_LOCAL_SECONDARY_INDEX_COUNT = 5;

export const CUSTOM_DDB_CFN_TYPE = 'Custom::AmplifyDynamoDBTable';

export interface AmplifyDynamoDBTableProps extends TableProps {
  customResourceServiceToken: string;
  allowDestructiveGraphqlSchemaUpdates?: boolean;
  replaceTableUponGsiUpdate?: boolean;
}
export class AmplifyDynamoDBTable extends Resource {
  public readonly encryptionKey?: kms.IKey;
  public readonly tableArn: string;
  public readonly tableName: string;
  public readonly tableStreamArn: string | undefined;
  public readonly tableFromAttr: ITable;

  private readonly customResourceServiceToken: string;
  private readonly table: CustomResource;
  private readonly keySchema = new Array<CfnTable.KeySchemaProperty>();
  private readonly attributeDefinitions = new Array<CfnTable.AttributeDefinitionProperty>();
  private readonly globalSecondaryIndexes = new Array<CfnTable.GlobalSecondaryIndexProperty>();
  private readonly localSecondaryIndexes = new Array<CfnTable.LocalSecondaryIndexProperty>();

  private readonly secondaryIndexSchemas = new Map<string, SchemaOptions>();
  private readonly nonKeyAttributes = new Set<string>();

  private readonly tablePartitionKey: Attribute;
  private readonly tableSortKey?: Attribute;

  private readonly billingMode: BillingMode;

  constructor(scope: Construct, id: string, props: AmplifyDynamoDBTableProps) {
    super(scope, id, {
      physicalName: props.tableName,
    });
    this.customResourceServiceToken = props.customResourceServiceToken;
    this.tableName = this.physicalName;
    const { sseSpecification, encryptionKey } = this.parseEncryption(props);

    let streamSpecification: CfnTable.StreamSpecificationProperty | undefined;
    this.billingMode = props.billingMode ?? BillingMode.PROVISIONED;
    if (props.stream) {
      streamSpecification = { streamViewType: props.stream };
    }

    this.validateProvisioning(props);

    // The 'Default' id is used for setting the construct default child
    // Refer https://docs.aws.amazon.com/cdk/api/v2/docs/constructs.Node.html#defaultchild
    this.table = new CustomResource(this, 'Default', {
      serviceToken: this.customResourceServiceToken,
      resourceType: CUSTOM_DDB_CFN_TYPE,
      properties: {
        tableName: this.tableName,
        attributeDefinitions: this.attributeDefinitions,
        keySchema: this.keySchema,
        globalSecondaryIndexes: Lazy.any({ produce: () => this.globalSecondaryIndexes }, { omitEmptyArray: true }),
        localSecondaryIndexes: Lazy.any({ produce: () => this.localSecondaryIndexes }, { omitEmptyArray: true }),
        pointInTimeRecoverySpecification:
          props.pointInTimeRecovery != null ? { pointInTimeRecoveryEnabled: props.pointInTimeRecovery } : undefined,
        billingMode: this.billingMode === BillingMode.PAY_PER_REQUEST ? this.billingMode : undefined,
        provisionedThroughput:
          this.billingMode === BillingMode.PAY_PER_REQUEST
            ? undefined
            : {
                readCapacityUnits: props.readCapacity || 5,
                writeCapacityUnits: props.writeCapacity || 5,
              },
        sseSpecification: sseSpecification,
        streamSpecification: streamSpecification,
        tableClass: props.tableClass,
        timeToLiveSpecification: props.timeToLiveAttribute ? { attributeName: props.timeToLiveAttribute, enabled: true } : undefined,
        deletionProtectionEnabled: props.deletionProtection,
        allowDestructiveGraphqlSchemaUpdates: props.allowDestructiveGraphqlSchemaUpdates ?? false,
        replaceTableUponGsiUpdate: props.replaceTableUponGsiUpdate ?? false,
      },
      removalPolicy: props.removalPolicy,
    });
    this.encryptionKey = encryptionKey;

    this.tableArn = this.table.getAttString('TableArn');
    this.tableStreamArn = streamSpecification ? this.table.getAttString('TableStreamArn') : undefined;

    this.tableFromAttr = Table.fromTableAttributes(scope, `CustomTable${id}`, {
      tableArn: this.tableArn,
      tableStreamArn: this.tableStreamArn,
    });

    this.addKey(props.partitionKey, HASH_KEY_TYPE);
    this.tablePartitionKey = props.partitionKey;

    if (props.sortKey) {
      this.addKey(props.sortKey, RANGE_KEY_TYPE);
      this.tableSortKey = props.sortKey;
    }

    this.node.addValidation({ validate: () => this.validateTable() });
  }

  /**
   * Add a global secondary index of table.
   *
   * @param props the property of global secondary index
   */
  public addGlobalSecondaryIndex(props: GlobalSecondaryIndexProps) {
    this.validateProvisioning(props);
    this.validateIndexName(props.indexName);

    // build key schema and projection for index
    const gsiKeySchema = this.buildIndexKeySchema(props.partitionKey, props.sortKey);
    const gsiProjection = this.buildIndexProjection(props);

    this.globalSecondaryIndexes.push({
      indexName: props.indexName,
      keySchema: gsiKeySchema,
      projection: gsiProjection,
      provisionedThroughput:
        this.billingMode === BillingMode.PAY_PER_REQUEST
          ? undefined
          : {
              readCapacityUnits: props.readCapacity || 5,
              writeCapacityUnits: props.writeCapacity || 5,
            },
    });

    this.secondaryIndexSchemas.set(props.indexName, {
      partitionKey: props.partitionKey,
      sortKey: props.sortKey,
    });
  }

  /**
   * Add a local secondary index of table.
   *
   * @param props the property of local secondary index
   */
  public addLocalSecondaryIndex(props: LocalSecondaryIndexProps) {
    // https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#limits-secondary-indexes
    if (this.localSecondaryIndexes.length >= MAX_LOCAL_SECONDARY_INDEX_COUNT) {
      throw new RangeError(`a maximum number of local secondary index per table is ${MAX_LOCAL_SECONDARY_INDEX_COUNT}`);
    }

    this.validateIndexName(props.indexName);

    // build key schema and projection for index
    const lsiKeySchema = this.buildIndexKeySchema(this.tablePartitionKey, props.sortKey);
    const lsiProjection = this.buildIndexProjection(props);

    this.localSecondaryIndexes.push({
      indexName: props.indexName,
      keySchema: lsiKeySchema,
      projection: lsiProjection,
    });

    this.secondaryIndexSchemas.set(props.indexName, {
      partitionKey: this.tablePartitionKey,
      sortKey: props.sortKey,
    });
  }

  /**
   * Get schema attributes of table or index.
   *
   * @returns Schema of table or index.
   */
  public schema(indexName?: string): SchemaOptions {
    if (!indexName) {
      return {
        partitionKey: this.tablePartitionKey,
        sortKey: this.tableSortKey,
      };
    }
    let schema = this.secondaryIndexSchemas.get(indexName);
    if (!schema) {
      throw new Error(`Cannot find schema for index: ${indexName}. Use 'addGlobalSecondaryIndex' or 'addLocalSecondaryIndex' to add index`);
    }
    return schema;
  }

  private addKey(attribute: Attribute, keyType: string) {
    const existingProp = this.findKey(keyType);
    if (existingProp) {
      throw new Error(`Unable to set ${attribute.name} as a ${keyType} key, because ${existingProp.attributeName} is a ${keyType} key`);
    }
    this.registerAttribute(attribute);
    this.keySchema.push({
      attributeName: attribute.name,
      keyType,
    });
    return this;
  }

  private findKey(keyType: string) {
    return this.keySchema.find((prop) => prop.keyType === keyType);
  }

  /**
   * Register the key attribute of table or secondary index to assemble attribute definitions of TableResourceProps.
   *
   * @param attribute the key attribute of table or secondary index
   */
  private registerAttribute(attribute: Attribute) {
    const { name, type } = attribute;
    const existingDef = this.attributeDefinitions.find((def) => def.attributeName === name);
    if (existingDef && existingDef.attributeType !== type) {
      throw new Error(`Unable to specify ${name} as ${type} because it was already defined as ${existingDef.attributeType}`);
    }
    if (!existingDef) {
      this.attributeDefinitions.push({
        attributeName: name,
        attributeType: type,
      });
    }
  }

  /**
   * Validate the table construct.
   *
   * @returns an array of validation error message
   */
  private validateTable(): string[] {
    const errors = new Array<string>();

    if (!this.tablePartitionKey) {
      errors.push('a partition key must be specified');
    }
    if (this.localSecondaryIndexes.length > 0 && !this.tableSortKey) {
      errors.push('a sort key of the table must be specified to add local secondary indexes');
    }

    return errors;
  }

  /**
   * Set up key properties and return the Table encryption property from the
   * user's configuration.
   */
  private parseEncryption(props: TableProps): { sseSpecification: CfnTableProps['sseSpecification']; encryptionKey?: kms.IKey } {
    let encryptionType = props.encryption;

    if (encryptionType === undefined) {
      encryptionType =
        props.encryptionKey != null
          ? // If there is a configured encryptionKey, the encryption is implicitly CUSTOMER_MANAGED
            TableEncryption.CUSTOMER_MANAGED
          : // Otherwise, if severSideEncryption is enabled, it's AWS_MANAGED; else undefined (do not set anything)
            TableEncryption.AWS_MANAGED;
    }

    if (encryptionType !== TableEncryption.CUSTOMER_MANAGED && props.encryptionKey) {
      throw new Error(
        '`encryptionKey cannot be specified unless encryption is set to TableEncryption.CUSTOMER_MANAGED (it was set to ${encryptionType})`',
      );
    }

    if (encryptionType === TableEncryption.CUSTOMER_MANAGED && props.replicationRegions) {
      throw new Error('TableEncryption.CUSTOMER_MANAGED is not supported by DynamoDB Global Tables (where replicationRegions was set)');
    }

    switch (encryptionType) {
      case TableEncryption.CUSTOMER_MANAGED:
        const encryptionKey =
          props.encryptionKey ??
          new kms.Key(this, 'Key', {
            description: `Customer-managed key auto-created for encrypting DynamoDB table at ${this.node.path}`,
            enableKeyRotation: true,
          });

        return {
          sseSpecification: { sseEnabled: true, kmsMasterKeyId: encryptionKey.keyArn, sseType: 'KMS' },
          encryptionKey,
        };

      case TableEncryption.AWS_MANAGED:
        // Not specifying "sseType: 'KMS'" here because it would cause phony changes to existing stacks.
        return { sseSpecification: { sseEnabled: true } };

      case TableEncryption.DEFAULT:
        return { sseSpecification: { sseEnabled: false } };

      case undefined:
        // Not specifying "sseEnabled: false" here because it would cause phony changes to existing stacks.
        return { sseSpecification: undefined };

      default:
        throw new Error(`Unexpected 'encryptionType': ${encryptionType}`);
    }
  }

  /**
   * Validate read and write capacity are not specified for on-demand tables (billing mode PAY_PER_REQUEST).
   *
   * @param props read and write capacity properties
   */
  private validateProvisioning(props: { readCapacity?: number; writeCapacity?: number }): void {
    if (this.billingMode === BillingMode.PAY_PER_REQUEST) {
      if (props.readCapacity !== undefined || props.writeCapacity !== undefined) {
        throw new Error('you cannot provision read and write capacity for a table with PAY_PER_REQUEST billing mode');
      }
    }
  }

  /**
   * Validate index name to check if a duplicate name already exists.
   *
   * @param indexName a name of global or local secondary index
   */
  private validateIndexName(indexName: string) {
    if (this.secondaryIndexSchemas.has(indexName)) {
      // a duplicate index name causes validation exception, status code 400, while trying to create CFN stack
      throw new Error(`a duplicate index name, ${indexName}, is not allowed`);
    }
  }

  /**
   * Validate non-key attributes by checking limits within secondary index, which may vary in future.
   *
   * @param nonKeyAttributes a list of non-key attribute names
   */
  private validateNonKeyAttributes(nonKeyAttributes: string[]) {
    if (this.nonKeyAttributes.size + nonKeyAttributes.length > 100) {
      // https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#limits-secondary-indexes
      throw new RangeError('a maximum number of nonKeyAttributes across all of secondary indexes is 100');
    }
    // store all non-key attributes
    nonKeyAttributes.forEach((att) => this.nonKeyAttributes.add(att));
  }

  private buildIndexKeySchema(partitionKey: Attribute, sortKey?: Attribute): CfnTable.KeySchemaProperty[] {
    this.registerAttribute(partitionKey);
    const indexKeySchema: CfnTable.KeySchemaProperty[] = [{ attributeName: partitionKey.name, keyType: HASH_KEY_TYPE }];

    if (sortKey) {
      this.registerAttribute(sortKey);
      indexKeySchema.push({ attributeName: sortKey.name, keyType: RANGE_KEY_TYPE });
    }

    return indexKeySchema;
  }

  private buildIndexProjection(props: SecondaryIndexProps): CfnTable.ProjectionProperty {
    if (props.projectionType === ProjectionType.INCLUDE && !props.nonKeyAttributes) {
      // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-dynamodb-projectionobject.html
      throw new Error(`non-key attributes should be specified when using ${ProjectionType.INCLUDE} projection type`);
    }

    if (props.projectionType !== ProjectionType.INCLUDE && props.nonKeyAttributes) {
      // this combination causes validation exception, status code 400, while trying to create CFN stack
      throw new Error(`non-key attributes should not be specified when not using ${ProjectionType.INCLUDE} projection type`);
    }

    if (props.nonKeyAttributes) {
      this.validateNonKeyAttributes(props.nonKeyAttributes);
    }

    return {
      projectionType: props.projectionType ?? ProjectionType.ALL,
      nonKeyAttributes: props.nonKeyAttributes ?? undefined,
    };
  }
}
