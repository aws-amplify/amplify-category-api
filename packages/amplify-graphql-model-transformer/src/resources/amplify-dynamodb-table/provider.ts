import { aws_lambda, aws_iam, Duration } from 'aws-cdk-lib';
import { Construct } from 'constructs';

/**
 * Lightweight construct wrapper for custom resource provider 'Custom::AmplifyDynamoDBTable'
 * Encapsulating the entry point (onEventHandler) within this `Provider` construct is necessary
 * to maintain a stable custom resource entry point ARN that is consistent with versions
 * previously deployed with the CDK's provider framework.
 *
 * IMPORTANT: The construct name and lambda function ID should not be changed.
 * These are components that make up the resolved ARN.
 * If you change anything in this construct, be sure to test schema modifications
 * against an already deployed stack using this custom resource.
 */
export class Provider extends Construct {
  public readonly onEventHandler: aws_lambda.Function;
  public readonly isCompleteHandler: aws_lambda.Function;
  public readonly serviceToken: string;

  constructor(scope: Construct, id: string, props: ProviderProps) {
    super(scope, id);

    this.onEventHandler = this.createFunction('onEvent', props.lambdaCode, props.onEventHandlerName, props.onEventRole);
    this.isCompleteHandler = this.createFunction('isComplete', props.lambdaCode, props.isCompleteHandlerName, props.isCompleteRole);
    this.serviceToken = this.onEventHandler.functionArn;
  }

  private createFunction = (entrypoint: string, code: aws_lambda.AssetCode, handler: string, role: aws_iam.IRole): aws_lambda.Function => {
    const fn = new aws_lambda.Function(this, `framework-${entrypoint}`, {
      code: code,
      description: `AmplifyManagedTable - ${entrypoint} (${this.node.path})`.slice(0, 256),
      runtime: aws_lambda.Runtime.NODEJS_18_X,
      handler,
      timeout: Duration.minutes(14),
      role,
    });

    return fn;
  };
}

export interface ProviderProps {
  readonly lambdaCode: aws_lambda.AssetCode;
  readonly onEventHandlerName: string;
  readonly onEventRole: aws_iam.IRole;
  readonly isCompleteHandlerName: string;
  readonly isCompleteRole: aws_iam.IRole;
}
