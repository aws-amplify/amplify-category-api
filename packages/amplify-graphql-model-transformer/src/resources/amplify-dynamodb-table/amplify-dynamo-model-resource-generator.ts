import * as cdk from 'aws-cdk-lib';
import { TransformerContextProvider } from '@aws-amplify/graphql-transformer-interfaces';
import { ModelResourceIDs, ResourceConstants } from 'graphql-transformer-common';
import { ObjectTypeDefinitionNode } from 'graphql';
import { setResourceName } from '@aws-amplify/graphql-transformer-core';
import { AttributeType, StreamViewType, TableEncryption } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

import { Duration, aws_iam, aws_lambda, custom_resources, aws_logs } from 'aws-cdk-lib';
import { DynamoModelResourceGenerator } from '../dynamo-model-resource-generator';
import * as path from 'path';
import { AmplifyDynamoDBTable } from './amplify-dynamodb-table-construct';

/**
 * AmplifyDynamoModelResourceGenerator is a subclass of DynamoModelResourceGenerator,
 * provisioning the DynamoDB tables with the custom resource instead of pre-defined DynamoDB table CFN template
 */

export const ITERATIVE_TABLE_STACK_NAME = 'AmplifyTableManager';
export class AmplifyDynamoModelResourceGenerator extends DynamoModelResourceGenerator {
  private customResourceServiceToken: string = '';
  private ddbManagerPolicy?: aws_iam.Policy;

  generateResources(ctx: TransformerContextProvider): void {
    if (!this.isEnabled()) {
      return;
    }

    if (this.isProvisioned()) {
      // add model related-parameters to the root stack
      const rootStack = cdk.Stack.of(ctx.stackManager.scope);
      this.createDynamoDBParameters(rootStack, false);

      const tableManagerStack = ctx.stackManager.getScopeFor('AmplifyTableCustomProvider', ITERATIVE_TABLE_STACK_NAME);
      this.createCustomProviderResource(tableManagerStack, ctx);
    }

    this.models.forEach((model) => {
      // This name is used by the mock functionality. Changing this can break mock.
      const tableBaseName = ctx.resourceHelper.getModelNameMapping(model!.name.value);
      const tableLogicalName = ModelResourceIDs.ModelTableResourceID(tableBaseName);
      const scope = ctx.stackManager.getScopeFor(tableLogicalName, tableBaseName);

      this.createModelTable(scope, model, ctx);
    });

    if (this.ddbManagerPolicy) {
      this.ddbManagerPolicy?.addStatements(
        new aws_iam.PolicyStatement({
          actions: [
            'dynamodb:CreateTable',
            'dynamodb:UpdateTable',
            'dynamodb:DeleteTable',
            'dynamodb:DescribeTable',
            'dynamodb:DescribeContinuousBackups',
            'dynamodb:DescribeTimeToLive',
            'dynamodb:UpdateContinuousBackups',
            'dynamodb:UpdateTimeToLive',
          ],
          resources: [
            cdk.Fn.sub('arn:aws:dynamodb:${AWS::Region}:${AWS::AccountId}:table/*-${apiId}-${envName}', {
              apiId: ctx.api.apiId,
              envName: ctx.synthParameters.amplifyEnvironmentName,
            }),
          ],
        }),
      );
    }

    this.generateResolvers(ctx);
  }

  protected createCustomProviderResource(scope: Construct, context: TransformerContextProvider): void {
    // Policy that grants access to Create/Update/Delete DynamoDB tables
    this.ddbManagerPolicy = new aws_iam.Policy(scope, 'CreateUpdateDeleteTablesPolicy');

    const lambdaCode = aws_lambda.Code.fromAsset(
      path.join(__dirname, '..', '..', '..', 'lib', 'resources', 'amplify-dynamodb-table', 'amplify-table-manager-lambda'),
    );

    // lambda that will handle DDB CFN events
    const gsiOnEventHandler = new aws_lambda.Function(scope, ResourceConstants.RESOURCES.TableManagerOnEventHandlerLogicalID, {
      runtime: aws_lambda.Runtime.NODEJS_18_X,
      code: lambdaCode,
      handler: 'amplify-table-manager-handler.onEvent',
      timeout: Duration.minutes(14),
    });

    // lambda that will poll for provisioning to complete
    const gsiIsCompleteHandler = new aws_lambda.Function(scope, ResourceConstants.RESOURCES.TableManagerIsCompleteHandlerLogicalID, {
      runtime: aws_lambda.Runtime.NODEJS_18_X,
      code: lambdaCode,
      handler: 'amplify-table-manager-handler.isComplete',
      timeout: Duration.minutes(14),
    });

    this.ddbManagerPolicy.attachToRole(gsiOnEventHandler.role!);
    this.ddbManagerPolicy.attachToRole(gsiIsCompleteHandler.role!);
    const customResourceProvider = new custom_resources.Provider(scope, ResourceConstants.RESOURCES.TableManagerCustomProviderLogicalID, {
      onEventHandler: gsiOnEventHandler,
      isCompleteHandler: gsiIsCompleteHandler,
      logRetention: aws_logs.RetentionDays.ONE_MONTH,
      queryInterval: Duration.seconds(30),
      totalTimeout: Duration.hours(2),
    });
    this.customResourceServiceToken = customResourceProvider.serviceToken;
  }

  protected createModelTable(scope: Construct, def: ObjectTypeDefinitionNode, context: TransformerContextProvider): void {
    const modelName = def!.name.value;
    const tableLogicalName = ModelResourceIDs.ModelTableResourceID(modelName);
    const tableName = context.resourceHelper.generateTableName(modelName);

    // Add parameters.
    const { readIops, writeIops, billingMode, pointInTimeRecovery, enableSSE } = this.createDynamoDBParameters(scope, true);

    // Add conditions.
    new cdk.CfnCondition(scope, ResourceConstants.CONDITIONS.HasEnvironmentParameter, {
      expression: cdk.Fn.conditionNot(cdk.Fn.conditionEquals(context.synthParameters.amplifyEnvironmentName, ResourceConstants.NONE)),
    });
    const useSSE = new cdk.CfnCondition(scope, ResourceConstants.CONDITIONS.ShouldUseServerSideEncryption, {
      expression: cdk.Fn.conditionEquals(enableSSE, 'true'),
    });
    const usePayPerRequestBilling = new cdk.CfnCondition(scope, ResourceConstants.CONDITIONS.ShouldUsePayPerRequestBilling, {
      expression: cdk.Fn.conditionEquals(billingMode, 'PAY_PER_REQUEST'),
    });
    const usePointInTimeRecovery = new cdk.CfnCondition(scope, ResourceConstants.CONDITIONS.ShouldUsePointInTimeRecovery, {
      expression: cdk.Fn.conditionEquals(pointInTimeRecovery, 'true'),
    });

    const removalPolicy = this.options.EnableDeletionProtection ? cdk.RemovalPolicy.RETAIN : cdk.RemovalPolicy.DESTROY;

    // TODO: The attribute of encryption and TTL should be added
    const table = new AmplifyDynamoDBTable(scope, `${tableLogicalName}`, {
      customResourceServiceToken: this.customResourceServiceToken,
      allowDestructiveGraphqlSchemaUpdates: context.transformParameters.allowDestructiveGraphqlSchemaUpdates,
      replaceTableUponGsiUpdate: context.transformParameters.replaceTableUponGsiUpdate,
      tableName,
      partitionKey: {
        name: 'id',
        type: AttributeType.STRING,
      },
      stream: StreamViewType.NEW_AND_OLD_IMAGES,
      encryption: TableEncryption.DEFAULT,
      removalPolicy,
      ...(context.isProjectUsingDataStore() ? { timeToLiveAttribute: '_ttl' } : undefined),
    });
    setResourceName(table, { name: modelName, setOnDefaultChild: false });

    // construct a wrapper around the custom table to allow normal CDK operations on top of it
    const tableRepresentative = table.tableFromAttr;
    setResourceName(tableRepresentative, { name: modelName, setOnDefaultChild: false });

    const cfnTable = table.node.defaultChild?.node.defaultChild as cdk.CfnCustomResource;
    setResourceName(cfnTable, { name: modelName, setOnDefaultChild: false });
    cfnTable.addPropertyOverride(
      'provisionedThroughput',
      cdk.Fn.conditionIf(usePayPerRequestBilling.logicalId, cdk.Fn.ref('AWS::NoValue'), {
        ReadCapacityUnits: readIops,
        WriteCapacityUnits: writeIops,
      }),
    );
    cfnTable.addPropertyOverride(
      'pointInTimeRecoverySpecification',
      cdk.Fn.conditionIf(usePointInTimeRecovery.logicalId, { PointInTimeRecoveryEnabled: true }, cdk.Fn.ref('AWS::NoValue')),
    );
    cfnTable.addPropertyOverride(
      'billingMode',
      cdk.Fn.conditionIf(usePayPerRequestBilling.logicalId, 'PAY_PER_REQUEST', cdk.Fn.ref('AWS::NoValue')).toString(),
    );
    cfnTable.addPropertyOverride('sseSpecification', {
      sseEnabled: cdk.Fn.conditionIf(useSSE.logicalId, true, false),
    });

    const streamArnOutputId = `GetAtt${ModelResourceIDs.ModelTableStreamArn(def!.name.value)}`;
    if (table.tableStreamArn) {
      new cdk.CfnOutput(cdk.Stack.of(scope), streamArnOutputId, {
        value: table.tableStreamArn,
        description: 'Your DynamoDB table StreamArn.',
        exportName: cdk.Fn.join(':', [context.api.apiId, 'GetAtt', tableLogicalName, 'StreamArn']),
      });
    }

    const tableNameOutputId = `GetAtt${tableLogicalName}Name`;
    new cdk.CfnOutput(cdk.Stack.of(scope), tableNameOutputId, {
      value: table.tableName,
      description: 'Your DynamoDB table name.',
      exportName: cdk.Fn.join(':', [context.api.apiId, 'GetAtt', tableLogicalName, 'Name']),
    });

    const role = this.createIAMRole(context, def, scope, tableName);
    const tableDataSourceLogicalName = `${def!.name.value}Table`;
    this.createModelTableDataSource(def, context, tableRepresentative, scope, role, tableDataSourceLogicalName);
  }
}
