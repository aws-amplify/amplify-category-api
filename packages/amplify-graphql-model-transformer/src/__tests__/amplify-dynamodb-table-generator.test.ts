import { testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import {
  DDB_AMPLIFY_MANAGED_DATASOURCE_STRATEGY,
  DDB_DEFAULT_DATASOURCE_STRATEGY,
  validateModelSchema,
} from '@aws-amplify/graphql-transformer-core';
import { parse } from 'graphql';
import { ModelTransformer } from '../graphql-model-transformer';
import { CUSTOM_DDB_CFN_TYPE } from '../resources/amplify-dynamodb-table/amplify-dynamodb-table-construct';
import { ITERATIVE_TABLE_STACK_NAME } from '../resources/amplify-dynamodb-table/amplify-dynamo-model-resource-generator';

describe('ModelTransformer:', () => {
  it('should successfully transform simple valid schema', async () => {
    const validSchema = `
      type Post @model {
          id: ID!
          title: String!
      }
      type Comment @model {
        id: ID!
        content: String
      }
    `;
    const out = testTransform({
      schema: validSchema,
      transformers: [new ModelTransformer()],
      dataSourceStrategies: {
        Comment: DDB_DEFAULT_DATASOURCE_STRATEGY,
        Post: DDB_AMPLIFY_MANAGED_DATASOURCE_STRATEGY,
      },
    });
    expect(out).toBeDefined();
    const amplifyTableManagerStack = out.stacks[ITERATIVE_TABLE_STACK_NAME];
    expect(amplifyTableManagerStack).toBeDefined();
    // DynamoDB manager policy should be generated correctly
    const ddbManagerPolicy = Object.values(amplifyTableManagerStack.Resources!)
      .filter((resource) => resource.Type === 'AWS::IAM::Role')
      .flatMap((role: any) => role.Properties.Policies)
      .filter((policies: any) => policies !== undefined)
      .reduce((acc, value) => acc.concat(value), [])
      .find((policy: any) => policy.PolicyName === 'CreateUpdateDeleteTablesPolicy');
    expect(ddbManagerPolicy).toBeDefined();
    expect(ddbManagerPolicy).toMatchSnapshot();
    // Post table resource should be generated within the custom table type
    const postStack = out.stacks['Post'];
    expect(postStack).toBeDefined();
    const postTable = postStack.Resources?.PostTable;
    expect(postTable).toBeDefined();
    expect(postTable.Type).toBe(CUSTOM_DDB_CFN_TYPE);
    // Comment table resource should be generated within the default CFN DynamoDB table
    const commentStack = out.stacks['Comment'];
    expect(commentStack).toBeDefined();
    const commentTable = commentStack.Resources?.CommentTable;
    expect(commentTable).toBeDefined();
    expect(commentTable.Type).toBe('AWS::DynamoDB::Table');
    validateModelSchema(parse(out.schema));

    // Outputs should contain a reference to the Arn to the entry point (onEventHandler)
    // of the provider for the AmplifyTableManager custom resource.
    // If any of these assertions should fail, it is likely caused by a change in the custom resource provider
    /** {@link Provider} */ // that caused the entry point ARN to change.
    // !! This will result in broken redeployments !!
    // Friends don't let friends mutate custom resource entry point ARNs.
    const outputs = amplifyTableManagerStack.Outputs!;
    expect(outputs).toBeDefined();
    const rootStackName = 'transformerrootstack';
    const tableManagerStackName = 'AmplifyTableManager';
    const onEventHandlerName = 'TableManagerCustomProviderframeworkonEvent';
    // See https://github.com/aws/aws-cdk/blob/6fdc4582f659549021a64a4d676fce12fc241715/packages/aws-cdk-lib/core/lib/stack.ts#L1288-L1333 for more information
    const entryPointKeyStableLogicalIdHash = 'F1C8BD67';
    const onEventHandlerStableLogicalIdHash = '1DFC2ECC';

    const entrypointArnOutputsKey = `${rootStackName}${tableManagerStackName}${onEventHandlerName}${entryPointKeyStableLogicalIdHash}Arn`;
    const entryPointOutput = outputs[entrypointArnOutputsKey];

    const onEventHandlerResourceName = `${onEventHandlerName}${onEventHandlerStableLogicalIdHash}`;
    expect(entryPointOutput).toBeDefined();
    expect(entryPointOutput['Value']['Fn::GetAtt']).toEqual([
      onEventHandlerResourceName /* TableManagerCustomProviderframeworkonEvent1DFC2ECC */,
      'Arn',
    ]);

    // Since we verified above that the ARN for this resource is included in the stack outputs,
    // we know that the resource itself exists in the stack. But better safe than sorry.
    const amplifyTableManagerResources = amplifyTableManagerStack.Resources;
    expect(amplifyTableManagerResources).toBeDefined();
    const onEventHandlerLambda = amplifyTableManagerResources![onEventHandlerResourceName];
    expect(onEventHandlerLambda).toBeDefined();
  });
});
