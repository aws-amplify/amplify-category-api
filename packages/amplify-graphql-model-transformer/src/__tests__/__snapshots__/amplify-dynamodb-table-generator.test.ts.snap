// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelTransformer: should successfully transform simple valid schema 1`] = `
Object {
  "PolicyDocument": Object {
    "Statement": Array [
      Object {
        "Action": Array [
          "dynamodb:CreateTable",
          "dynamodb:UpdateTable",
          "dynamodb:DeleteTable",
          "dynamodb:DescribeTable",
          "dynamodb:DescribeContinuousBackups",
          "dynamodb:DescribeTimeToLive",
          "dynamodb:UpdateContinuousBackups",
          "dynamodb:UpdateTimeToLive",
        ],
        "Effect": "Allow",
        "Resource": Object {
          "Fn::Sub": Array [
            "arn:aws:dynamodb:\${AWS::Region}:\${AWS::AccountId}:table/*-\${apiId}-\${envName}",
            Object {
              "apiId": Object {
                "Ref": "referencetotransformerrootstackGraphQLAPI20497F53ApiId",
              },
              "envName": Object {
                "Ref": "referencetotransformerrootstackenv10C5A902Ref",
              },
            },
          ],
        },
      },
    ],
    "Version": "2012-10-17",
  },
  "PolicyName": "CreateUpdateDeleteTablesPolicy",
}
`;
